master/buildbot/www/hooks/bitbucket.py

```diff
@@ -12,24 +12,161 @@
-    """Catch a POST request from BitBucket and start a build process
+
+_HEADER_CT = 'Content-Type'
+_HEADER_EVENT = 'X-Event-Key'
+HOSTED_GIT_URL = "git@bitbucket.org"
+
+
+class BitbucketEventHandler(object):
+
+    def __init__(self, codebase=None):
+        self._codebase = codebase
+
+    def process(self, request):
+        """
+        Process a webhook event request.
+
+        See example on https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/hooks
+        for list of all the different events.
+        """
+        payload = self._get_payload(request)
+        event_type = request.getHeader(_HEADER_EVENT)
+        log.msg("Processing event %s: %r" % (_HEADER_EVENT, event_type,), logLevel=logging.DEBUG)
+        event_type = event_type.replace(":", "_")
+        handler = getattr(self, 'handle_%s' % event_type, None)
+
+        if handler is None:
+            raise ValueError('Unknown event: %r' % (event_type,))
+
+        return handler(payload)
+
+    def _get_payload(self, request):
+        content = request.content.read()
+        content_type = request.getHeader(_HEADER_CT)
+
+        if content_type == 'application/json':
+            payload = json.loads(content)
+        elif content_type == 'application/x-www-form-urlencoded':
+            payload = json.loads(request.args['payload'][0])
+        else:
+            raise ValueError('Unknown content type: %r' % (content_type,))
+
+        log.msg("Payload: %r" % payload, logLevel=logging.DEBUG)
+
+        return payload
+
+    def handle_repo_push(self, payload):
+        repo = payload['repository']
+        project = repo['name']
+        push = payload['push']
+        payload['ref'] = push['changes'][0]["new"]["name"]
+        repo_url = "%s:%s.git" % (HOSTED_GIT_URL, payload['repository']['full_name'])
+        changes = self._process_change(payload, repo, repo_url, project)
+        log.msg("Received %d changes from Bitbucket" % len(changes))
+        return changes, 'git'
+
+    def handle_pullrequest_updated(self, payload):
+        return self.handle_pullrequest(payload, 'updated_on')
+
+    def handle_pullrequest_created(self, payload):
+        return self.handle_pullrequest(payload, 'created_on')
+
+    def _process_change(self, payload, repo, repo_url, project):
+        """
+        Consumes the JSON as a python object and returns the changes.
+
+        :arguments:
+            payload
+                Python Object that represents the JSON sent by Bitbucket Webhook event.
+        """
+        changes = []
+
+        for change_ in payload['push']['changes']:
+            if change_['new'] is None:
+                # Branch deleted
+                continue
+            branch = change_['new']['name']
+            when_timestamp = dateparse(change_['new']['target']['date'])
+            for commit in change_['commits']:
+                log.msg("New revision: %s" % commit['hash'][:8])
```

Comment:
Dont think it is a good idea to add a log for every change, as the db will contain that change with more information